subsys/bluetooth/host/classic/hfp_ag.c

```diff
@@ -385,256 +357,1118 @@ static int hfp_ag_send_data(struct bt_hfp_ag *ag, bt_hfp_ag_tx_cb_t cb, void *us
+			codec_ids |= BIT(codec);
+		}
+	}
+
+	if (!(codec_ids & BIT(BT_HFP_AG_CODEC_CVSD))) {
+		return -EOPNOTSUPP;
+	}
+
+	hfp_ag_lock(ag);
+	ag->hf_codec_ids = codec_ids;
+	ag->selected_codec_id = 0;
+	hfp_ag_unlock(ag);
+
+	atomic_set_bit(ag->flags, BT_HFP_AG_CODEC_CHANGED);
+
+	if (atomic_test_and_clear_bit(ag->flags, BT_HFP_AG_CODEC_CONN)) {
+		/* Codec connection is ended. It needs to be restarted. */
+		LOG_DBG("Codec connection is ended. It needs to be restarted.");
+		if (bt_ag && bt_ag->codec_negotiate) {
+			bt_ag->codec_negotiate(ag, -EAGAIN);
+		}
+	}
+
+	if (bt_ag && bt_ag->codec) {
+		bt_ag->codec(ag, ag->hf_codec_ids);
+	}
+
+	return 0;
+}
+
+static int bt_hfp_ag_cind_handler(struct bt_hfp_ag *ag, struct net_buf *buf)
+{
+	int err;
+	bool inquiry_status = true;
+
+	if (is_char(buf, '=')) {
+		inquiry_status = false;
+	}
+
+	if (!is_char(buf, '?')) {
+		return -ENOTSUP;
+	}
+
+	if (!is_char(buf, '\r')) {
+		return -ENOTSUP;
+	}
+
+	if (!inquiry_status) {
+		err = hfp_ag_send_data(
+			ag, NULL, NULL,
+			"\r\n+CIND:(\"%s\",(%d%s%d)),(\"%s\",(%d%s%d)),(\"%s\",(%d%s%d)),(\"%s\",(%"
+			"d%s%d)),(\"%s\",(%d%s%d)),(\"%s\",(%d%s%d)),(\"%s\",(%d%s%d))\r\n",
+			ag_ind[BT_HFP_AG_SERVICE_IND].name, ag_ind[BT_HFP_AG_SERVICE_IND].min,
+			ag_ind[BT_HFP_AG_SERVICE_IND].connector, ag_ind[BT_HFP_AG_SERVICE_IND].max,
+			ag_ind[BT_HFP_AG_CALL_IND].name, ag_ind[BT_HFP_AG_CALL_IND].min,
+			ag_ind[BT_HFP_AG_CALL_IND].connector, ag_ind[BT_HFP_AG_CALL_IND].max,
+			ag_ind[BT_HFP_AG_CALL_SETUP_IND].name, ag_ind[BT_HFP_AG_CALL_SETUP_IND].min,
+			ag_ind[BT_HFP_AG_CALL_SETUP_IND].connector,
+			ag_ind[BT_HFP_AG_CALL_SETUP_IND].max, ag_ind[BT_HFP_AG_CALL_HELD_IND].name,
+			ag_ind[BT_HFP_AG_CALL_HELD_IND].min,
+			ag_ind[BT_HFP_AG_CALL_HELD_IND].connector,
+			ag_ind[BT_HFP_AG_CALL_HELD_IND].max, ag_ind[BT_HFP_AG_SIGNAL_IND].name,
+			ag_ind[BT_HFP_AG_SIGNAL_IND].min, ag_ind[BT_HFP_AG_SIGNAL_IND].connector,
+			ag_ind[BT_HFP_AG_SIGNAL_IND].max, ag_ind[BT_HFP_AG_ROAM_IND].name,
+			ag_ind[BT_HFP_AG_ROAM_IND].min, ag_ind[BT_HFP_AG_ROAM_IND].connector,
+			ag_ind[BT_HFP_AG_ROAM_IND].max, ag_ind[BT_HFP_AG_BATTERY_IND].name,
+			ag_ind[BT_HFP_AG_BATTERY_IND].min, ag_ind[BT_HFP_AG_BATTERY_IND].connector,
+			ag_ind[BT_HFP_AG_BATTERY_IND].max);
+	} else {
+		err = hfp_ag_send_data(ag, NULL, NULL, "\r\n+CIND:%d,%d,%d,%d,%d,%d,%d\r\n",
+					 ag->indicator_value[BT_HFP_AG_SERVICE_IND],
+					 ag->indicator_value[BT_HFP_AG_CALL_IND],
+					 ag->indicator_value[BT_HFP_AG_CALL_SETUP_IND],
+					 ag->indicator_value[BT_HFP_AG_CALL_HELD_IND],
+					 ag->indicator_value[BT_HFP_AG_SIGNAL_IND],
+					 ag->indicator_value[BT_HFP_AG_ROAM_IND],
+					 ag->indicator_value[BT_HFP_AG_BATTERY_IND]);
+	}
+
+	return err;
+}
+
+static void bt_hfp_ag_set_in_band_ring(struct bt_hfp_ag *ag, void *user_data)
+{
+	bool is_inband_ringtone;
+
+	hfp_ag_lock(ag);
+	is_inband_ringtone = (ag->ag_features & BT_HFP_AG_FEATURE_INBAND_RINGTONE) ? true : false;
+	hfp_ag_unlock(ag);
```

Comment:
Does the locking really give you anything here?

Reply:
Yes. The feature only be updated during the SLC connection.

Reply:
Updated.